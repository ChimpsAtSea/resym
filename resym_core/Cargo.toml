[package]
name = "resym_core"
version.workspace = true
edition = "2021"
rust-version = "1.75"
exclude = [
    "tests/data/*",
    "tests/snapshots/*",
]

[features]
default = ["rayon", "http"]

rayon = ["dep:rayon", "dep:num_cpus", "dashmap/rayon"]
http = ["dep:ehttp", "dep:url"]

[dependencies]
pdb = { git = "https://github.com/ergrelet/pdb", branch = "0.8.0-resym" }
thiserror = "1.0"
log = "0.4"
rayon = { version = "1.5", optional = true }
num_cpus = { version = "1.13", optional = true }
crossbeam-channel = "0.5"
dashmap = { version = "5.2" }
regex = "1.5"
similar = "2.1"
ehttp = { version = "0.2", optional = true }
url = { version = "2.3", optional = true }

# Web:
[target.'cfg(target_arch = "wasm32")'.dependencies]
wasm_thread = { version = "0.2", features = ["es_modules"] }
instant = { version = "0.1", features = ["wasm-bindgen"] }

[dev-dependencies]
insta = "1.14"
