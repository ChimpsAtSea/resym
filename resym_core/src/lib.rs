pub mod backend;
pub mod diffing;
mod error;
pub mod frontend;
pub mod pdb_file;
pub mod pdb_types;
pub mod syntax_highlighting;

pub use error::*;

const PKG_VERSION: &str = env!("CARGO_PKG_VERSION");

/// Macro used to switch between iterators depending on rayon's availability
#[macro_export]
#[cfg(not(feature = "rayon"))]
macro_rules! cond_par_iter {
    ($expression:expr) => {
        $expression.iter()
    };
}
#[macro_export]
#[cfg(feature = "rayon")]
macro_rules! cond_par_iter {
    ($expression:expr) => {
        $expression.par_iter()
    };
}

/// Macro used to switch between functions depending on rayon's availability
#[macro_export]
#[cfg(not(feature = "rayon"))]
macro_rules! cond_sort_by {
    ($expression:expr, $($x:tt)*) => {
        $expression.sort_by($($x)*)
    };
}
#[macro_export]
#[cfg(feature = "rayon")]
macro_rules! cond_sort_by {
    ($expression:expr, $($x:tt)*) => {
        $expression.par_sort_by($($x)*)
    };
}
