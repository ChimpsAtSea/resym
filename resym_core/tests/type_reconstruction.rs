use std::path::Path;

use resym_core::{pdb_file::PdbFile, pdb_types::PrimitiveReconstructionFlavor};

const TEST_PDB_FILE_PATH: &str = "tests/data/test.pdb";
const TEST_CASES: &[&str] = &[
    "resym_test::PrimitiveTypesTest",
    "resym_test::ArrayTest",
    "resym_test::BitFieldsTest1",
    "resym_test::BitFieldsTest2",
    "resym_test::UnionTest",
    "resym_test::StructTest",
    "resym_test::EnumTest1",
    "resym_test::EnumTest2",
    "resym_test::StructUnnamedUdtTest1",
    "resym_test::StructUnnamedUdtTest2",
    "resym_test::StructUnnamedUdtTest3",
    "resym_test::UnionUnnamedUdtTest1",
    "resym_test::PureVirtualClassSpecialized",
    "resym_test::InterfaceImplClass",
    "resym_test::SpecializedInterfaceImplClass",
    "resym_test::ClassWithRefsAndStaticsTest",
];

#[test]
fn test_type_reconstruction_portable_access_specifiers() {
    test_type_reconstruction_internal(
        "type_reconstruction_portable_access_specifiers",
        PrimitiveReconstructionFlavor::Portable,
        false,
        true,
    );
}

fn test_type_reconstruction_internal(
    test_name: &str,
    primitives_flavor: PrimitiveReconstructionFlavor,
    reconstruct_dependencies: bool,
    print_access_specifiers: bool,
) {
    let pdb_file = PdbFile::load_from_file(Path::new(TEST_PDB_FILE_PATH)).expect("load test.pdb");
    for (i, test_case_type_name) in TEST_CASES.iter().enumerate() {
        let reconstructed_type = pdb_file
            .reconstruct_type_by_name(
                test_case_type_name,
                primitives_flavor,
                reconstruct_dependencies,
                print_access_specifiers,
            )
            .expect(format!("reconstruct type: {}", test_case_type_name).as_str());

        let snapshot_name = format!("{}-{}", test_name, i);
        insta::assert_snapshot!(snapshot_name, reconstructed_type);
    }
}
