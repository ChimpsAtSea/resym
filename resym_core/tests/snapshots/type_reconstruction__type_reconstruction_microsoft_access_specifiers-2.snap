---
source: resym_core/tests/type_reconstruction.rs
expression: reconstructed_type
---

struct resym_test::BitFieldsTest1 { /* Size=0x4 */
  union {
    /* 0x0000 */ public: ULONG b1 : 1;
    /* 0x0000 */ public: ULONG b2 : 1;
    /* 0x0000 */ public: ULONG b3 : 30;
  };
};

